staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go

```diff
@@ -187,14 +190,12 @@ func (c *controllerCommon) AttachDisk(isManagedDisk bool, diskName, diskURI stri
 		return -1, fmt.Errorf("failed to get azure instance id for node %q (%v)", nodeName, err)
 	}
 
-	c.vmLockMap.LockEntry(strings.ToLower(string(nodeName)))
```

Comment:
the lock is actually for L202(AttachDisk) operation, below function is sending a data disk list to Azure,  so if there are 2 disk attach operation in parallel, then 2 different data disk list would be generated, and should be last win, that's a limitation from Azure CRP, we have to add a lock when there is UpdateVM operation(here is AttachDisk operation)

https://github.com/kubernetes/kubernetes/blob/6d01c5a58996d1619ac049c2b3077274299eb2d0/staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_vmss.go#L66-L73

Reply:
Oh, I did not know that about azure API.
/close

We need to wait for Azure to speed up their attach times then. Batching could help, but I am not sure about how much - in my case, a volume was being attached and another attach call started 1 minute after that. Even with quite generous batching, there would be probably two separate Azure API calls that must be sequential. Unfortunately, the first call took >5 minutes.

Reply:
single disk fast attach/detach would costs ~10s, that would definitely help, batch may help a little IMO. I will let you know ASAP when that fast attach/detach feature is in public preview. Thanks.